### DISCUSSION TITLE
Testing lifecycle of layer that performs async computation

### DISCUSSION THREAD

**Discussion by cneumann** [visgl/deck.gl#5594] [Q&A] [answered]
Hello,
I have a (few) custom `CompositeLayer`s that compute values from their `data` prop and pass those values to their sub layers for rendering. The computation can take a little time, so it happens asynchronously and when it completes the composite layers internal state is updated.

```javascript
updateState({props, oldProps}) {
  if (props.data !== oldProps.data) {
     // compute new derived data, store in local state when done
     computeDerivedData(props.data).then((derivedData) => this.setState({derivedData: derivedData});
  }
}

renderLayers() {
  return [
    new SomeLayer(this.getSubLayerProps({
        // use derived data from local state
        data: this.state.derivedData
      })
    )
  ]
}
```
This all works fine, but now I want to write (lifecycle) tests for these layers and I'm not sure how to best do this.

If I use `testLayer` (from @deck.gl/test-utils) the problem is that by the time `onAfterUpdate` is called the async computation is not finished and I cannot make assertions about the new state yet.

I also tried using `testLayerAsync` but that appears to hang indefinitely and after a quick look at its implementation I believe it is because it is targeted at layers that take a url as their `data` prop and perform data loading internally (because [here](https://github.com/visgl/deck.gl/blob/04f38c5bbecefb8d54a82577eaa7ebf40ab68c1c/modules/test-utils/src/lifecycle-test.js#L91) it checks `layer.isLoaded`).

I think a variation of `testLayerAsync` that supports a callback `layerReady` which can return true once the layer has finished all async operations and is used in place of the `layer.isLoaded` check in the above while loop could solve my problem, but perhaps there is another way that I'm missing?

**Comment by Pessimistress** [comment, marked as answer]
Just implement your own `isLoaded`:

```js
get isLoaded() {
  // return asyncComputationCompleted;
}
```

**Reply by cneumann** [reply]
Thanks for the suggestion, trying it quickly it seems to solve my problem :smile:
